### PR TITLE
StepFunctions: Migration to String Expressions

### PR DESCRIPTION
**Pull request by MEPalma**
## Motivation
Currently the SFN internpreter v2 represents string literals though a `keyword_or_string` sub-production or through specialised terminal token types. This approach lacked an intermediary representation for string carrying executable expressions (jsonpaths, context objects, intrinsic functions, jsonata expressions, variable sampling), leading to some logic duplication as each declaration often having to implement the evaluation logic of these strings in varying context. These changes introduce the `string_expression` sub-production and a few logic child productions. Most importantly, all its sub productions are preprocessed as the expected evaluation component for the specific string expression they represent. The result is a more delcarative grammar, which ensures the correct string expressions are placed in the intended productions, with the productions seemingly having to just delegate the string expression provided, or apply the relevant logic overrides.

## Changes
- introduced the `string_expression` concept
- introduced reusable string expression evaluation component
- updated the grammar to be `more delcarative` in the expected string expressions for each sub-production
- applied relevant changes for this migration

### CODE REVIEW

File: localstack-core/localstack/services/stepfunctions/asl/static_analyser/variable_references_static_analyser.py [hunk 1]
```diff
@@ -67,36 +61,35 @@ def _put_variable_name(self, variable_name: VariableName) -> None:
...
+    #     if maybe_string_var is not None:
+    #         self._extract_variable_references_from_string_var(terminal_node=maybe_string_var)
+    #
+    #     maybe_intrinsic_function = Antlr4Utils.is_terminal(
+    #         pt=node, token_type=ASLLexer.STRINGINTRINSICFUNC
+    #     )
+    #     if maybe_intrinsic_function is not None:
+    #         self._extract_variable_references_from_intrinsic_function(
+    #             terminal_node=maybe_intrinsic_function
+    #         )
```
Comment (gregfurman) [inline localstack-core/localstack/services/stepfunctions/asl/static_analyser/variable_references_static_analyser.py:95]: I imagine this should be removed?
```suggestion
```

File: localstack-core/localstack/services/stepfunctions/asl/component/state/state_wait/wait_function/seconds_path.py [hunk 2]
```diff
@@ -56,21 +58,11 @@ def _validate_seconds_value(self, env: Environment, seconds: Any):
...
-    def _get_wait_seconds(self, env: Environment) -> int:
-        self.variable_sample.eval(env=env)
+        self.string_sampler.eval(env=env)
         seconds = env.stack.pop()
         self._validate_seconds_value(env=env, seconds=seconds)
         return seconds
+        # inp = env.stack[-1]
+        # seconds = extract_json(self.path, inp)
+        # self._validate_seconds_value(env=env, seconds=seconds)
+        # return seconds
```
Comment (gregfurman) [inline localstack-core/localstack/services/stepfunctions/asl/component/state/state_wait/wait_function/seconds_path.py:68]: Remove:

```suggestion
```

**Review by gregfurman — COMMENTED** [review]
Left some questions. Happy to approve when those comments are removed 😄 

Also, very happy with this approach!

File: localstack-core/localstack/services/stepfunctions/asl/component/common/string/string_expression.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        # TODO: this could be optimised by sampling only those invoked.
+        variable_declarations: VariableDeclarations = env.variable_store.get_variable_declarations()
+        variable_declarations_list.append(variable_declarations)
+
+        rich_jsonata_expression: JSONataExpression = compose_jsonata_expression(
+            final_jsonata_expression=self.expression,
+            variable_declarations_list=variable_declarations_list,
+        )
+        result = eval_jsonata_expression(rich_jsonata_expression)
+        env.stack.append(result)
```
[the remark this answers is not in the document]
Comment (MEPalma) [reply localstack-core/localstack/services/stepfunctions/asl/component/common/string/string_expression.py:111]: Yes indeed, in this PR I am simply migrating the current logic to reusable string expression. This is definately a peformance aspect we will improve on

File: localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py [hunk 4]
```diff
@@ -137,7 +118,13 @@ def _validate_value(self, env: Environment, value: int) -> None:
...
-        inp = env.stack[-1]
-        max_items = extract_json(self.path, inp)
+        self.string_sampler.eval(env=env)
+        max_items = env.stack.pop()
+        if isinstance(max_items, str):
+            try:
+                max_items = int(max_items)
+            except Exception:
+                # Pass incorrect type forward for validation and error reporting
+                pass
```
[the remark this answers is not in the document]
Comment (MEPalma) [reply localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py:128]: In this migration i left the max items computation value within the _get_value routine, of whcih output is then passed to the _clip_value routine. Zero values are remapped to the default max value.
However, yes I agree, in general, expressions's output validation does need increasing in parity, not only wrt MaxItems!

**Review by gregfurman — APPROVED** [review]
LGTM!

**Pull request merged**